source/Utility.js

```diff
@@ -1,18 +1,22 @@
 exports.hasCollision = function hasCollision(a, b) {
-    var x1 = a.position.x
-    var y1 = a.position.y
-    var w1 = a.width
-    var h1 = a.height
+    if ( a == null || b == null) {
+        return false
+    } else {
+        var x1 = a.position.x
+        var y1 = a.position.y
+        var w1 = a.width
+        var h1 = a.height
 
-    var x2 = b.position.x
-    var y2 = b.position.y
-    var w2 = b.width
-    var h2 = b.height
+        var x2 = b.position.x
+        var y2 = b.position.y
+        var w2 = b.width
+        var h2 = b.height
 
-    if(x1 < x2 + w2 && x1 + w1 > x2
-    && y1 < y2 + h2 && h1 + y1 > y2) {
-        return true
-    } else {
-        return false
+        if(x1 < x2 + w2 && x1 + w1 > x2
+        && y1 < y2 + h2 && h1 + y1 > y2) {
+            return true
+        } else {
+            return false
+        }
```

Comment:
We could, but there won't be much shared code, honestly. hasCollision could go there, sure... but controllers and inputs are going to only apply to Junkerships (and are probably going to need a class of their own once the time comes for multiplayer support) and statics can reside in Reference or Utility as appropriate. As for particles, we've got the explosion effect, but I imagine we're going to want the Junkership deaths to be a bit more dramatic than the Trashbot explosions. The parallax effect starfield doesn't behave like the other Entities, so they wouldn't be appropriate for tying to this either. Are there other particle effects you had in mind?

If Javascript had something more akin to an Interface, I'd say go with it, but nearly everything we need is in Pixi.Sprite anyway. Seems overkill to add an abstract class (which looks to be a bit clunky in Javascript, though a bit better in ES6) for what is going to mostly get overridden.